storage/remote/queue_manager.go

```diff
@@ -1504,6 +1553,22 @@ func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.Ti
 	// without causing a memory leak, and it has the nice effect of not propagating any
 	// parameters for sendSamplesWithBackoff/3.
 	attemptStore := func(try int) error {
+		if s.qm.cfg.SampleAgeLimit != 0 {
+			// This will filter out old samples during retries.
+			req, _, err := buildWriteRequest(
```

Comment:
What is going on here?  Why it is it before the start of the span?

Reply:
The idea here is to filter out batches that might get old during multiple attempts of sending samples. There's no reason of putting before the start of the span (probably missed it). Do we want to send the span even if errors happen?

Reply:
Why does the comment say "during retries" when it is called every time?
In the typical case, is it repeating the work done at the start of `sendSamplesWithBackoff`?  That doesn't seem appealing.

Reply:
Indeed. The 'during retries' refers to subsequent calls to attemptStore.

The original idea here is that some of the timeseries might be crossing the 'too old' threshold during retries, so they should be filtered out on these subsequent request by re-building the write request when an age limit has been set.

Given the output of `buildWriteRequest` is the request itself, I believe it would be too expensive to unmarshall, filter out and marshal back and the only alternative here is that we don't filter on subsequent requests; samples are dropped as too old only when the request is built, leading to more samples slipping through to be rejected by the remote write.

Do you feel that's a preferred route?

Reply:
Disagree this is "the only alternative".
You have a `try` parameter which could be used to implement the "during retries" idea.
Or you could pass in an "oldest sample" time which could be checked against the current time to see if anything will be filtered.

Reply:
+1 for `Or you could pass in an "oldest sample" time which could be checked against the current time to see if anything will be filtered.`, if we know the oldest timestamp in the prev. write request is still not old enough to be filtered we can skip building the write request an additional time.